### DISCUSSION TITLE
task-local static for tests

### DISCUSSION THREAD

**Discussion by Urhengulas** [tokio-rs/tokio#3169] [Q&A]
Hi ^^

### context
For some server I am building (which uses tokio) we are using a global variable for the database connection:
```rust
use diesel::{PgConnection, r2d2::{ConnectionManager, Pool}};
use once_cell::sync::OnceCell;

static DB: OnceCell<Pool<ConnectionManager<PgConnection>>> = OnceCell::new();
```

### problem
This generally works very well, but now we are adding some tests (using `#[tokio::test]`) and hit a problem there:
All of our tests shall run against a fresh, separate database to avoid side-effects. But since they all share the same static variable this currently doesn't work!

### solution ideas
1. The easiest solution would be to set `worker_threads = 1`, but this is not desired.
2. My idea would be to have a static variable which is local to the test. Since a test is a tokio-task _(as far as I understand)_ it would be a "task-local static variable".

Is there something like that? Or do you seem some other (better?) solution or workaround?

Merci beaucoup  <3

**Comment by davidbarsky** [comment]
Tokio does have support for task locals https://docs.rs/tokio/0.3.4/tokio/task/struct.LocalKey.html, which you can use to scope these connections.

However, I'll note that `#[tokio::test]` isn't a single task; it's a runtime spun up _for that test_ that can an arbitrary number of tasks.

**Reply by Urhengulas** [reply]
And I guess there isn't a thing such as a "runtime-local static"?

**Comment by Darksonn** [comment]
It seems to me that the best solution would be to simply not store it in a global variable. Then you can create a separate one in each test. I don't think there's any way to avoid multiple tests from using the same global. There are task-locals, sure, but those don't transfer across calls to `tokio::spawn`.

As for `#[tokio::test]`, that macro will in fact create a new separate single-threaded executor for every test. It's not just multiple tasks on one runtime.

**Reply by Urhengulas** [reply]
This would mean that I always need to explicitly pass my database connection to the functions using it, which is obviously doable, but having it in the static is kinda "nicer"..

**Reply by Darksonn** [reply]
Well yes, but if it's a global, you can only have one.

**Reply by Urhengulas** [reply]
> Well yes, but if it's a global, you can only have one.

Sure, but my problem is, that it works very well and is very convenient for the actual application, since I only need one database connection for it. 
Sacrificing this for the test feels kinda bad, but probably there is no easy way around this...

**Reply by dmgolembiowski** [reply]
@Urhengulas I ran into a similar problem, and this is what I came up with: https://github.com/dmgolembiowski/leptos-ecommerce/blob/0912c16ee577e2c7676ef8e205b67b43ce2a9a42/common/src/lib.rs#L21
